src/OSPSuite.R/Services/ConcurrentSimulationRunner.cs

```diff
@@ -80,35 +111,52 @@ public void Clear()
+      public ConcurrentSimulationResults[] RunConcurrently()
       {
          //Currently we only allow for running simulations or simulation batches, but not both
          if (_simulationBatches.Count > 0 && _simulations.Count > 0)
-            //Temporal Exception. We should allow for mixing both use cases but we need to discuss first
             throw new OSPSuiteException(Error.InvalidMixOfSimulationAndSimulationBatch);
 
          _cancellationTokenSource = new CancellationTokenSource();
          if (_simulations.Count > 0)
          { 
-            var results = _concurrentManager.RunAsync(
+            return _concurrentManager.RunAsync(
                NumberOfCores,
                _cancellationTokenSource.Token,
                _simulations,
                runSimulation
-            ).Result;
-            return _simulations.Select(s => results[s]).ToArray();
+            ).Result.Values.ToArray();
          }
 
          if (_simulationBatches.Count > 0)
-            return null;
+         {
+            return _concurrentManager.RunAsync<SimulationBatchRunOptions, ConcurrentSimulationResults>(
+               NumberOfCores,
+               _cancellationTokenSource.Token,
+               _simulationBatches.SelectMany(sb => sb.SimulationBatchRunValues.Select(rv => new SimulationBatchRunOptions() { Simulation = sb.Simulation, SimulationBatchOptions = sb.SimulationBatchOptions, SimulationBatchRunValues = rv })).ToList(),
+               runSimulationBatch
+            ).Result.Values.ToArray();
+         }
 
-         return Array.Empty<SimulationResults>();
+         return Array.Empty<ConcurrentSimulationResults>();
       }
 
-      private Task<SimulationResults> runSimulation(int coreIndex, CancellationToken cancellationToken, IModelCoreSimulation simulation)
+      private async Task<ConcurrentSimulationResults> runSimulation(int coreIndex, CancellationToken cancellationToken, IModelCoreSimulation simulation)
       {
          //We want a new instance every time that's why we are not injecting SimulationRunner in constructor
-         return Api.GetSimulationRunner().RunAsync(new SimulationRunArgs {Simulation = simulation, SimulationRunOptions = SimulationRunOptions});
+         return new ConcurrentSimulationResults(
+            simulation.Id,
+            await Api.GetSimulationRunner().RunAsync(new SimulationRunArgs { Simulation = simulation, SimulationRunOptions = SimulationRunOptions })
+         );
+      }
+
+      private async Task<ConcurrentSimulationResults> runSimulationBatch(int coreIndex, CancellationToken cancellationToken, SimulationBatchRunOptions simulationBatchWithOptions)
+      {
+         return new ConcurrentSimulationResults(
+            simulationBatchWithOptions.Simulation.Id,
+            await Api.GetSimulationBatchFactory().Create(simulationBatchWithOptions.Simulation, simulationBatchWithOptions.SimulationBatchOptions).RunAsync(simulationBatchWithOptions.SimulationBatchRunValues),
```

Comment:
in fact this is not ok here. You are going to create one instance for each instances and thus creating the SImModel multiple times. This is not ok. The Batch should be created only once and then updated

Reply:
If we share the same copy of the SimModel then the simulation runs will interfere with each other. I think we discussed this before. It is actually similar to why we create a new SimulationRunner with Api.GetSimulationRunner() on every runSimulation method. You could try to keep a reference to the same SimulationBatch and run with different values concurrently and see the errors.

Reply:
I disagree with this. There is absolutely not advantage in that case to using SmulationBatch and we can remove it altogether.

Reply:
You mean that the time necessary for cloning the Simulation is actually higher than the gain of running concurrently I guess. You might be right in that. Removing it altogether is a bit too drastic I would say. What if we would create the SimulationBatch objects when they are first used (on each thread for better performance) and the references are kept until the ConcurrentSimulationRunner is disposed? That could be an intermediate solution, right? This way the first time you RunAsync will be a bit longer but still will worth if you plan to run the simulations a bunch of times. We could also take the initiative and create the SimulationBatches on a concurrent way as well when the method AddSimulationBatchOption is invoked. What do you think?

Reply:
>You could try to keep a reference to the same SimulationBatch and run with different values concurrently and see the errors

The problem you had before was that you were using the same instance of the SimModelBatch. here you should have  multiple instances. 
Otherwise, and I might be wrong, you will create one SimModelInstance for EACH simulationrun. The gain is then very very limited

Reply:
>What if we would create the SimulationBatch objects when they are first used (on each thread for better performance) and the references are kept until the ConcurrentSimulationRunner is disposed?

This is what I thought you would do after our discussion on Monday. The best of both world. // run and optimized for speed

Reply:
Ok, then I did not understand correct at some point but now that I fully understand, I will do the changes next. I will let you know when the implementation is ready. Thanks for the discussion... was a good feedback for me.

Reply:
Awesome, We can discuss about this tomorrow if you want

Reply:
Sure, I have just committed a possible solution where I keep a list of objects that needs the SimulationBatch to be created. These SimulationBatch objects are only created once before they will be used for the first time and using also the ConcurrencyManager. The only problem is that the cores are not used as efficient as possible. The current solution is consistent in the sense that the class SimulationWithBatchOptions handles the Batches which makes sense since this is the class where you add new SimulationBatchRunValues so you can easily keep track of it. The problem is that if we would move the logic for initializing the batches to ConcurrentSimulationRunner, then the cores could be used in a more performant way but if we do so, I think the solution will be more fragile. If the R caller uses the wrong combination of calls, it could crash since the logic is not well encapsulated. I would prefer to go for robustness now but we could discuss and maybe a couple of external eyes could help me see what I am lacking now. I hope this could be a good compromise for now.

Reply:
I also just improved the solution in terms of performance without making the information public in R. I did mark the data internal which is not very usual. Maybe we could discuss in our meeting in a few hours.

Reply:
Updated to what we discussed today